### ISSUE TITLE
If testng can support @ test if it fails, if it fails to execute afterClass, it executes afterClass if @test runs successfully？

### ISSUE THREAD

**Issue by yanyan0305** [testng-team/testng#2119]
### TestNG Version
> Note: only the latest version is supported
testng 6.13.1
### Expected behavior
If `@Test` fails to run, `@AfterClass` cleanup environment is not performed.

### Actual behavior
If `@Test` fails to run, execute `@AfterClass`

### Is the issue reproductible on runner?

- [ ] Shell
- [ ] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [ ] IntelliJ
- [ ] NetBeans

### Test case sample
> Please, share the test case (as small as possible) which shows the issue

**Comment by krmahadevan** [comment]
I retried this with TestNG `6.14.3` and its not reproducible. 

Note: The current latest released version of TestNG is `7.0.0-beta7`. Please try this again with that.
If this is still a problem, please share a sample that can be used to recreate the issue.

**Comment by yanyan0305** [comment]
How to set the @Test execution failure does not run @afterClass?The following example：
public class TestCase {
    @BeforeClass
    private void setUp(){
        System.out.println("---begin to exec @beforeClass.");
    }
    @Test
    private void test() {
        System.out.println("---begin to exec @Test");
        Assert.fail("---exec @test fail!");
    }
    @AfterClass
    private void tearDown(){
        System.out.println("---begin to exec @afterClass.");
    }
results：
---begin to exec @beforeClass.
---begin to exec @Test
java.lang.AssertionError: ---exec @test fail!
...............................
---begin to exec @afterClass.
Here @afterClass is bound to execute，I want to execute @afterClass, only if @test execution succeeds. @afterclass is not run if @ test execution fails.

**Comment by krmahadevan** [comment]
@yanyan0305 - That cannot be done with `@AfterClass`. What you are asking for is conditional execution of `@AfterClass` config annotation based on test status. That changes the meaning of this annotation and I don't think that can be supported.

Please explain what exactly are you trying to achieve as an email on the testng-users google forum with a proper sample.

@juherr @cbeust - Thoughts ?

**Comment by juherr** [comment]
@krmahadevan Do you reproduce the issue?

`@AfterClass` is not supposed to be run if a test fails unless you have `alwaysRun=true`.

> `@AfterClass`: The annotated method will be run after all the test methods in the current class have been run. 

> alwaysRun: 
> For after methods (afterSuite, afterClass, ...): If set to true, this configuration method will be run even if one or more methods invoked previously failed or was skipped.

https://testng.org/doc/documentation-main.html#annotations

**Comment by krmahadevan** [comment]
@juherr 

> @AfterClass is not supposed to be run if a test fails unless you have alwaysRun=true.

I have gone back all the way till `6.8.8` and what you say is not what TestNG is doing. TestNG executes the `@AfterClass` method all the time. `alwaysRun=true` attribute doesn't seem to have anything to do with the `@Test` method's result. I believe that the `alwaysRun` attribute for an `@AfterXX` method indicates if a teardown configuration should be executed, even if there was a failure in the setup configuration method.. so in this case, `@AfterClass(alwaysRun=true)` would force TestNG to run this method, even if there was a failure in `@BeforeClass` annotated method.

Please see https://github.com/cbeust/testng/blob/c4f803ebba7bfadb558fb0b8a7dc9c4e92d82a57/src/main/java/org/testng/internal/ConfigInvoker.java#L234-L266

Fixing this issue will change the behavior of TestNG which has existed for over 5 years.

**Comment by juherr** [comment]
@krmahadevan Yes, maybe but it is not how I understand the documentation, do you? 

In fact, I think the issue exists for a long time because

> alwaysRun configuration methods:  if a test fails, any
configuration methods annotated with this boolean will always be run,
regardless of whether something wrong happened during the tests.

https://beust.com/weblog/2005/08/08/announcing-testng-2-5/

And I didn't find in the changelog a real reason for the behavior change.

But I agree with the backward compatible argument too.

I didn't find any tests that check that feature, is it really the same behavior for all the configuration methods?

At least, I think we should add test cases (maybe in 
https://github.com/cbeust/testng/tree/master/src/test/java/test/alwaysrun) and improve the documentation.

Thoughts?

**Comment by krmahadevan** [comment]
@juherr 

> is it really the same behavior for all the configuration methods?

I don't know. I haven't tested it out yet.

> At least, I think we should add test cases (maybe in https://github.com/cbeust/testng/tree/master/src/test/java/test/alwaysrun) and improve the documentation.

Sure we can. Just to ensure that we are on the same page.

* Add unit tests to test the current behaviour.
* Modify documentation/javadocs etc., to explicitly explain the current behaviour.

Is this what you are expecting ?

**Comment by ken-p** [comment]
>I want to execute @AfterClass, only if @Test execution succeeds.

@yanyan0305 simple to just check the result and execute conditionally:

```java
@AfterClass
public void after(ITestResult result) {
        boolean success = result.isSuccess();
        if (success) {
            //...
        }
}
```

**Comment by juherr** [comment]
AI triage verdict: **still relevant** as an enhancement request.

The current behavior is directly established by `ConfigInvoker`: configuration eligibility is suppressed by prior **configuration** failures (`@BeforeSuite`/`@BeforeClass`/etc.), but a normal `@Test` failure is not recorded as a configuration failure. Consequently a plain `@AfterClass` remains eligible after a failed test method. This also matches the historical investigation in this issue, which found the behavior going back at least to 6.8.8.

Current `test.alwaysrun.AlwaysRunTest` coverage distinguishes teardown behavior after failed setup/configuration methods, but I found no focused current test for the exact scenario here: successful `@BeforeClass`, failing `@Test`, then plain `@AfterClass`. I found no `#2119`/`GITHUB-2119` source, test, `CHANGES.txt`, or commit reference and no PR implementing conditional `@AfterClass` execution. Open PR #3453 is related only in that it clarifies `alwaysRun` semantics around **configuration failures**; it does not implement this request.

The historical reproducer was **not executed** during this triage. The current behavior is confirmed structurally, but changing it would alter a long-standing lifecycle contract and needs an explicit compatibility/design decision.

Next step: first add a focused contract test with a successful `@BeforeClass`, a deliberately failing `@Test`, and an `@AfterClass` that records whether it ran. Then decide whether TestNG should preserve unconditional class teardown or add a separate opt-in mechanism for conditional teardown; only after that decision should production behavior change.
